src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs

```diff
@@ -0,0 +1,333 @@
+                    {
+                        yield return func;
+                    }
+                }
+            }
+        }
+
+        internal TexlFunction Append(TexlFunction function) => Add(function);
+
+        internal TexlFunction Add(TexlFunction function)
+        {
+            if (function == null)
+            {
+                throw new ArgumentNullException($"{nameof(function)} cannot be null", nameof(function));
+            }
+
+            var fList = WithNameInternal(function.Name);
+
+            if (fList.Any())
+            {
+                if (fList.Contains(function))
+                {
+                    throw new ArgumentException($"Function {function.Name} is already part of core or extra functions");
+                }
+
+                fList.Add(function);
+            }
+            else
+            {
+                _functions.Add(function.Name, new List<TexlFunction>() { function });
+            }
+
+            var fInvariantList = WithInvariantNameInternal(function.LocaleInvariantName);
+
+            if (fInvariantList.Any())
+            {
+                if (fInvariantList.Contains(function))
+                {
+                    throw new ArgumentException($"Function {function.Name} is already part of core or extra functions (invariant)");
+                }
+
+                fInvariantList.Add(function);
+            }
+            else
+            {
+                _functionsInvariant.Add(function.LocaleInvariantName, new List<TexlFunction>() { function });
+            }
+
+            var fnsList = WithNamespaceInternal(function.Namespace);
+
+            if (fnsList.Any())
+            {
+                if (fnsList.Contains(function))
+                {
+                    throw new ArgumentException($"Function {function.Name} is already part of core or extra functions (namespace)");
+                }
+
+                fnsList.Add(function);
+            }
+            else
+            {
+                _namespaces.Add(function.Namespace, new List<TexlFunction>() { function });
+            }
+
+            _enums.AddRange(function.GetRequiredEnumNames());
+
+            _count++;
+
+            return function;
+        }        
+
+        internal TexlFunctionSet Add(TexlFunctionSet functionSet)
+        {
+            if (functionSet == null)
+            {
+                throw new ArgumentNullException($"{nameof(functionSet)} cannot be null", nameof(functionSet));
+            }
+
+            if (_count == 0)
+            {
+                _functions = new Dictionary<string, List<TexlFunction>>(functionSet._functions);
+                _functionsInvariant = new Dictionary<string, List<TexlFunction>>(functionSet._functionsInvariant);
+                _namespaces = new Dictionary<DPath, List<TexlFunction>>(functionSet._namespaces);
+                _enums = new List<string>(functionSet._enums);
+                _count = functionSet._count;
+            }
```

Comment:
Can we just omit this block, and let the case below handle it?

Reply:
this is a valuable perf improvement vs. several loops

Reply:
this would be worth adding as a comment